tests/Tests/Protocols/Wishbone.hs

```diff
@@ -0,0 +1,160 @@
+{-# LANGUAGE NumericUnderscores  #-}
+{-# LANGUAGE RecordWildCards     #-}
+-- The exhaustiveness-checker doesn't work well with ViewPatterns
+{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
+module Tests.Protocols.Wishbone where
+
+
+import           Clash.Prelude                  hiding ((&&))
+
+import           Clash.Hedgehog.Sized.BitVector
+
+import           Hedgehog
+
+import           Hedgehog.Gen                   as Gen
+
+import           Hedgehog.Range                 as Range
+
+import           Protocols
+import           Protocols.Hedgehog             (defExpectOptions)
+import           Protocols.Wishbone
+import           Protocols.Wishbone.Hedgehog
+
+-- tasty
+import           Control.DeepSeq                (NFData)
+import           Data.Either                    (isLeft)
+import           Prelude                        hiding (undefined)
+import           Test.Tasty
+import           Test.Tasty.Hedgehog            (HedgehogTestLimit (HedgehogTestLimit))
+import           Test.Tasty.Hedgehog.Extra      (testProperty)
+import           Test.Tasty.TH                  (testGroupGenerator)
+
+smallInt :: Range Int
+smallInt = Range.linear 0 10
+
+genSmallInt :: Gen Int
+genSmallInt = Gen.integral smallInt
+
+genData :: Gen a -> Gen [a]
+genData = Gen.list (Range.linear 0 150)
+
+genWishboneTransfer
+  :: (KnownNat addressWidth)
+  => Gen a
+  -> Gen (WishboneMasterRequest addressWidth a)
+genWishboneTransfer genA = Gen.choice
+  [ Read <$> genDefinedBitVector
+  , Write <$> genDefinedBitVector <*> genA ]
+
+--
+-- 'id' circuit
+--
+
+idWriteWbSt
+  :: forall a dom addressWidth
+  . (BitPack a, NFDataX a)
+  => Circuit (Wishbone dom 'Standard addressWidth a) ()
+idWriteWbSt = Circuit go
+  where
+    go (m2s, ()) = (reply <$> m2s, ())
+
+    reply WishboneM2S{..}
+      | busCycle && strobe && writeEnable =
+          (emptyWishboneS2M @a) { acknowledge = True
+                                , readData = writeData }
+      | busCycle && strobe                = emptyWishboneS2M { acknowledge = True }
+      | otherwise                         = emptyWishboneS2M
+
+idWriteStModel
+  :: (NFData a, Eq a, ShowX a)
+  => WishboneMasterRequest addressWidth a
+  -> WishboneS2M a
+  -> () -- ^ pure state
+  -> Either String ()
+idWriteStModel (Read _)    s@WishboneS2M{..} ()
+  | acknowledge && isLeft (hasX readData) = Right ()
+  | otherwise                             =
+      Left $ "Read should have been acknowledged with no DAT " <> showX s
+idWriteStModel (Write _ a) s@WishboneS2M{..} ()
+  | acknowledge && hasX readData == Right a = Right ()
+  | otherwise                               =
+      Left $ "Write should have been acknowledged with write-data as DAT " <> showX s
+
+prop_idWriteSt_model :: Property
+prop_idWriteSt_model =
+  wishbonePropWithModel @System
+    defExpectOptions
+    idWriteStModel
+    idWriteWbSt
+    (genData $ genWishboneTransfer @10 genSmallInt)
+    ()
+
+--
+-- memory element circuit
+--
+
```

Comment:
This was meant to be a spacing/grouping comment, because the memory element circuit testing has the actual wishbone circuit and the pure model function as well. It felt a little "squished" with all the different functions for the different components coming one after the other.